### PR TITLE
feat: create a branch from origin owner, repo, branch name, and an optional primary branch

### PR DESCRIPTION
**Pull request by TomKristie**
This PR implements the basic functionality of creating a branch and has basic mock test.

Given:
- an owner
- a repository
- a branch name
- an optional primary branch name

Correctly create a new GitHub reference based off the primary branch HEAD SHA.
On successful reference creation, the GitHub API (wrapped by Octokit) returns the base SHA for new commits to be applied on top of.

On GitHub API error, log error and re-throw the error. On primary branch not found error, throw that error.

This PR also groups repository domain data together, and reverts accidental `.gitattribute` and `synth.metadata` changes

Towards #19

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -1,3 +1,3 @@
 *.ts text eol=lf
-*.js test eol=lf
+*.js text eol=lf
```
Comment (TomKristie) [inline .gitattributes:2]: reverted change from last PR

File: synth.metadata [hunk 2]
```diff
@@ -4,7 +4,7 @@
       "git": {
         "name": ".",
         "remote": "https://github.com/googleapis/code-suggester.git",
-        "sha": "900b4920694588f53b1fc1894a806d6d13056ba7"
+        "sha": "c62c02fb99f26f13b796f976ad1f7a2a4384660d"
```
Comment (TomKristie) [inline synth.metadata:7]: reverted change from last PR

File: .gitattributes (hunk 1 above)
Comment (chingor13) [reply .gitattributes:2]: I would separate this into a separate PR. There shouldn't be any conflicts with this feature branch.

File: src/github-handler/branch-handler.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+function createRef(branchName: string) {
+  return refPrefix + branchName;
+}
+
+/**
+ * Return the length of the GitHub reference prefix
+ */
+function refPrefixLen() {
+  return refPrefix.length;
+}
```
Comment (chingor13) [inline src/github-handler/branch-handler.ts:33]: If you really need this (seems a bit weird to need), make it a constant and export the value directly.

File: src/github-handler/branch-handler.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {Logger, Octokit, RepoDomain} from '../types';
+
+const refPrefix = 'refs/heads/';
+const masterBranch = 'master';
```
Comment (chingor13) [inline src/github-handler/branch-handler.ts:18]: Let's future-proof this:
```suggestion
const DEFAULT_PRIMARY_BRANCH = 'master';
```

File: src/types/index.ts [hunk 5]
```diff
@@ -20,6 +20,14 @@ interface Files {
   [index: string]: string;
 }
 
+/**
+ * The domain of a repository
+ */
+interface RepoDomain {
+  repo: string;
+  owner: string;
+}
```
Comment (chingor13) [inline src/types/index.ts:29]: Another conceptual model that we may want to consider is repo+owner+branch name. I think we could possibly have another interface:
```ts

interface RepoBranch extends RepoDomain {
  branch: string;
}
```

File: src/github-handler/branch-handler.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+ * @param {string} primaryBranch the name of the primary branch. Default is master
+ * @returns {Promise<string>} the base SHA for subsequent commits to be based off for the origin branch
+ */
+async function branch(
+  logger: Logger,
+  octokit: Octokit,
+  origin: RepoDomain,
+  originBranchName: string,
+  primaryBranch: string = masterBranch
+): Promise<string> {
```
Comment (chingor13) [inline src/github-handler/branch-handler.ts:79]: Just looking at the argument names, it's a bit hard to tell where we're branch from and to.

If we have the RepoBranch type, we might make this:
```ts
async function branch(
  logger: Logger,
  octokit: Octokit,
  from: RepoBranch,
  to: RepoBranch
): Promise<string>
```

Comment (TomKristie) [reply src/github-handler/branch-handler.ts:79]: There's no from/to relationship. The branch is always created off of the origin with a particular branch name. The primary branch is to get the latest SHA.

File: src/github-handler/branch-handler.ts (hunk 3 above)
Comment (TomKristie) [reply src/github-handler/branch-handler.ts:33]: This actually is not longer used so this will be deleted.

File: src/types/index.ts (hunk 5 above)
Comment (TomKristie) [reply src/types/index.ts:29]: Let's table this discussion for when we want to support multiple upstream origin repositories

File: src/github-handler/branch-handler.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  branch: string
+): Promise<string> {
+  const branchData = (
+    await octokit.repos.getBranch({
+      owner: origin.owner,
+      repo: origin.repo,
+      branch,
+    })
+  ).data;
+  logger.info('Successfully found primary branch HEAD sha.');
```
Comment (chingor13) [inline src/github-handler/branch-handler.ts:50]: nit: generally, we want to add contextual information to any log statements that we plan to keep (like the actual sha in this case)

File: src/github-handler/branch-handler.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      owner: origin.owner,
+      repo: origin.repo,
+      ref: createRef(name),
+      sha: baseSHA,
+    });
+    logger.info('Successfully created branch');
+    return baseSHA;
+  } catch (err) {
+    logger.error('Error when creating branch');
+    throw Error(err.toString());
```
Comment (chingor13) [inline src/github-handler/branch-handler.ts:84]: Can we rethrow the same error? Wrapping it doesn't seem to buy us anything.

File: test/branch.ts [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+  };
+
+  describe('Octokit list branch function', () => {
+    it('is passed the correct parameters', async () => {
+      // setup
+      const getBranchStub = sinon
+        .stub(octokit.repos, 'getBranch')
+        .resolves(branchResponse);
+      // // tests
+      await getBranchHead(logger, octokit, origin, 'master');
```
Comment (chingor13) [inline test/branch.ts:49]: We should still capture the return value and assert its value is correctly parsed.

File: test/branch.ts [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+        },
+      };
+      const getBranchStub = sinon
+        .stub(octokit.repos, 'getBranch')
+        .resolves(branchResponse);
+      const createRefStub = sinon
+        .stub(octokit.git, 'createRef')
+        .resolves(createRefResponse);
+      // tests
+      await branch(logger, octokit, origin, branchName, 'master');
```
Comment (chingor13) [inline test/branch.ts:89]: Same here, let's capture the return value and assert its value.

File: src/github-handler/branch-handler.ts [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {Logger, Octokit, RepoDomain} from '../types';
+
+const refPrefix = 'refs/heads/';
```
Comment (chingor13) [inline src/github-handler/branch-handler.ts:17]: nit: for a constant, let's make it `REF_PREFIX`

File: src/github-handler/branch-handler.ts [hunk 12]
```diff
@@ -71,17 +73,19 @@ async function branch(
...
+        ref: createRef(name),
+        sha: baseSHA,
+      })
+    ).data;
+    logger.info(`Successfully created branch at ${refData.url}`);
     return baseSHA;
   } catch (err) {
     logger.error('Error when creating branch');
-    throw Error(err.toString());
+    throw Error(err);
```
Comment (chingor13) [inline src/github-handler/branch-handler.ts:88]: For this one will this work?
```suggestion
    throw err;
```

Comment (TomKristie) [reply src/github-handler/branch-handler.ts:88]: Thanks for revisiting this. I realized this actually does work but not with the current tests.

**Pull request merged**